src/diffCheck/segmentation/DFSegmentation.cc

```diff
@@ -6,95 +6,227 @@
-            cilantroPointCloud.normals, 
+            cilantroPointCloud->normals,
             normalThresholdDegree*M_PI/180.0f);
 
-            // Segment the point cloud
-            cilantro::ConnectedComponentExtraction3f<> segmenter(cilantroPointCloud.points);
+            cilantro::ConnectedComponentExtraction3f<> segmenter(cilantroPointCloud->points);
             segmenter.segment(neighborhood, similarityEvaluator, minClusterSize);
 
             auto clusterToPointMap = segmenter.getClusterToPointIndicesMap();
             int nSegments = segmenter.getNumberOfClusters();
 
-            // Get the segments
             for (int indice = 0; indice<nSegments; indice++)
             {
                 std::shared_ptr<geometry::DFPointCloud> segment = std::make_shared<geometry::DFPointCloud>();
                 for (auto pointIndice : clusterToPointMap[indice])
                 {
-                    Eigen::Vector3d point = cilantroPointCloud.points.col(pointIndice).cast<double>();
-                    Eigen::Vector3d normal = cilantroPointCloud.normals.col(pointIndice).cast<double>();
+                    Eigen::Vector3d point = cilantroPointCloud->points.col(pointIndice).cast<double>();
+                    Eigen::Vector3d normal = cilantroPointCloud->normals.col(pointIndice).cast<double>();
                     segment->Points.push_back(point);
                     segment->Normals.push_back(normal);
+                    if (cilantroPointCloud->hasColors())
+                    {
+                        Eigen::Vector3d color = cilantroPointCloud->colors.col(pointIndice).cast<double>();
+                        segment->Colors.push_back(color);
+                    }
                 }
+                if (colorClusters)
+                    segment->ApplyColor(Eigen::Vector3d::Random());
                 segments.push_back(segment);
             }
         }
-        
 
         return segments;
     }
+
+    std::shared_ptr<geometry::DFPointCloud> DFSegmentation::AssociateClusters(
+        std::vector<std::shared_ptr<geometry::DFMesh>> referenceMesh,
+        std::vector<std::shared_ptr<geometry::DFPointCloud>> &clusters,
+        double associationThreshold)
+    {
+        std::shared_ptr<geometry::DFPointCloud> unifiedPointCloud = std::make_shared<geometry::DFPointCloud>();
+        std::vector<std::shared_ptr<geometry::DFPointCloud>> segmentsRemainder;
+
+        // iterate through the mesh faces given as function argument
+        for (std::shared_ptr<diffCheck::geometry::DFMesh> face : referenceMesh)
+        {
+            std::shared_ptr<geometry::DFPointCloud> correspondingSegment;
+            std::shared_ptr<open3d::geometry::TriangleMesh> o3dFace;
+            o3dFace = face->Cvt2O3DTriangleMesh();
+
+            // Getting the center of the mesh face
+            Eigen::Vector3d faceCenter;
+            open3d::geometry::OrientedBoundingBox orientedBoundingBox = o3dFace->GetMinimalOrientedBoundingBox();
```

Comment:
Having a OBB for getting  the centroid of a mesh face seems a bit brittle, maybe a simpler mean of vertices would do the trick?

Reply:
Moreover, if the o3dFace is in the x, y, or z plane, it is a OBB with volume = 0, which freaks out open3d. I have fixed this and use now the o3dFace->GetCenter() method (which I should have done from the beginning)